### PR TITLE
Fix incorrect behavior of record codec code generator

### PR DESCRIPTION
**Pull request by ahydrax**
Hi there!
This fixes https://github.com/dotnet/orleans/issues/8286.

The problem is that if derived record has the same properties as base record e.g.
```
  [GenerateSerializer]
    [Id(8286)]
    public abstract record TestType2(string TestString)
    {
        [GenerateSerializer]
        public sealed record TestType2A(string TestString) : TestType2(TestString);

        [GenerateSerializer]
        public sealed record TestType2B(string TestString) : TestType2(TestString);
    }
```
then the code generator will generate extra call (even if we're not writing any fields).
```
public void Serialize<TBufferWriter>(ref global::Orleans.Serialization.Buffers.Writer<TBufferWriter> writer, global::OrleansBugRepro.Interfaces.TestType2.TestType2B instance)
    where TBufferWriter : global::System.Buffers.IBufferWriter<byte>
{
    _baseTypeSerializer.Serialize(ref writer, instance);
    writer.WriteEndBase();
    writer.WriteEndBase();  // <-- redundant call
}
```

So I assumed that it only happens with records as it have more things inside than regular class hierarchy (primary constructors, overriden ToString/Equals, etc) and then I tried to do the same thing with class hierarchy:
```
[Immutable, GenerateSerializer, Alias("TestType2")]
public abstract class TestType2
{
    public string? TestString { get; }

    protected TestType2(string? testString)
    {
        TestString = testString;
    }

    [Immutable, GenerateSerializer, Alias("TestType2A")]
    public sealed class TestType2A : TestType2
    {
        public TestType2A(string? testString) : base(testString)
        {
        }
    }

    [Immutable, GenerateSerializer, Alias("TestType2B")]
    public sealed class TestType2B : TestType2
    {
        public TestType2B(string? testString) : base(testString)
        {
        }
    }
}
```
and we're okay with that. The code is generated properly and tests are pass!
```
public void Serialize<TBufferWriter>(ref global::Orleans.Serialization.Buffers.Writer<TBufferWriter> writer, global::OrleansBugRepro.Interfaces.TestType2.TestType2B instance)
            where TBufferWriter : global::System.Buffers.IBufferWriter<byte>
        {
            _baseTypeSerializer.Serialize(ref writer, instance);
            writer.WriteEndBase();
            // <-- no redundant call here
        }
```

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8288)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=8288) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=8288) before we can accept your contribution.<br/><br/>:x: ahydrax [sign now](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=8288)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/orleans?pullRequest=8288) it.</sub>

**Comment by ahydrax** [comment]
Hi @ReubenBond , I mentioned you because you're assigned to this https://github.com/dotnet/orleans/issues/8286

**Comment by ReubenBond** [comment]
Hi @ahydrax, thank you for taking a look at this! I only see changes under the test/ directory in the commits here.

**Comment by ahydrax** [comment]
@ReubenBond could you review that? I would also like to add more unit test to cover polymorphic cases. If have any ideas about test case I would appreciate that.

**Comment by ahydrax** [comment]
I found a similar issue https://github.com/dotnet/orleans/issues/8275

File: src/Orleans.CodeGenerator/SerializerGenerator.cs [hunk 1]
```diff
@@ -341,7 +342,10 @@ private static BaseCodecFieldDescription GetBaseTypeField(ISerializableTypeDescr
             if (type.SupportsPrimaryConstructorParameters)
             {
                 AddSerializationMembers(type, serializerFields, members.Where(m => m.IsPrimaryConstructorParameter), libraryTypes, writerParam, instanceParam, previousFieldIdVar, body);
-                body.Add(ExpressionStatement(InvocationExpression(writerParam.Member("WriteEndBase"), ArgumentList())));
+                if (body.LastOrDefault() != writeEndBaseInvocationExpression)
```
Comment (ReubenBond) [inline src/Orleans.CodeGenerator/SerializerGenerator.cs:345]: This seems a little fragile - if some unrelated code is inserted in between then this check would break

**Comment by ReubenBond** [comment]
I should have mentioned, but I have been working on this in response to #8275, too. I opened a PR for my approach: https://github.com/dotnet/orleans/pull/8289

**Comment by ahydrax** [comment]
Since you're more familiar with the codebase I'm sure that your approach will be better than mine. I'll close this PR so I have only one question: any ETA?
